### PR TITLE
Bt 16.1 - Create a Cities View

### PR DESCRIPTION
**Pull request by peytonshetler**
## Summary

Created a cities view (basically a duplicate of the states view).  

#### Characteristics

- [X] Feature completed
- [ ] Experimental
- [X] UI affected (screen shots included)

#### Testing

- [ ] included
- [X] Not Included 

Screen Shot 2019-10-03 at 2 34 41 PM

Screen Shot 2019-10-03 at 2 34 52 PM

### CODE REVIEW

File: Been-There/Been-There/Controller/LandingScreen/UI+LandingScreen.swift [hunk 1]
```diff
@@ -135,19 +135,21 @@ extension LandingScreenViewController {
     }
     
     func setupMainTitle() {
-        mainTitle.translatesAutoresizingMaskIntoConstraints = false
         mainTitle.text = "Been There"
-        mainTitle.font = UIFont.mainFontItalic(size: 48)
+        //mainTitle.font = UIFont.mainFontItalic(size: 48)
+        mainTitle.font = UIFont.preferredFont(forTextStyle: .largeTitle)
```
Comment (JZDesign) [inline Been-There/Been-There/Controller/LandingScreen/UI+LandingScreen.swift]: # 👍

File: Been-There/Been-There/Controller/LandingScreen/UI+LandingScreen.swift [hunk 2]
```diff
@@ -135,19 +135,21 @@ extension LandingScreenViewController {
...
-        mainTitle.centerXAnchor.constraint(equalTo: topContainer.centerXAnchor).isActive = true
-        mainTitle.centerYAnchor.constraint(equalTo: topContainer.centerYAnchor, constant: -30).isActive = true
-        mainTitle.widthAnchor.constraint(equalTo: topContainer.widthAnchor).isActive = true
-        mainTitle.heightAnchor.constraint(equalTo: topContainer.heightAnchor, multiplier: 1/4).isActive = true
+        mainTitle
+            .centerX(topContainer.centerXAnchor)
+            .centerY(topContainer.centerYAnchor, with: -30)
+            .width(topContainer.widthAnchor)
+            .height(topContainer.heightAnchor, withMultiplier: 1/4)
+            .build()
```
Comment (JZDesign) [inline Been-There/Been-There/Controller/LandingScreen/UI+LandingScreen.swift]: # 👍

**Comment by JZDesign** [comment]
# Duplication

> ## (basically a duplicate of the states view).

This means we could abstract a view and reuse it instead of copy and paste it. Let's say you pull out all of your States view layout and setup logic to a new controller like `CardCollectionViewController` or `BTCollectionViewController`. The only functions you would really need are one to populate the data and to override `cellForRowAt` for both the States View and Cities View if they inherit from that new controller. 

All of the layout and styling would be done, you could implement you new controllers with less than 1/4 the code. 

```swift

class BTCollectionViewController: UIViewController, UICollectionViewDelegate, UICollectionViewDelegateFlowLayout, UICollectionViewDataSource {

    var locations: [Location] = []
    let topContainer = UIView()
    let bottomContainer = UIView()
    let mainTitle = UILabel()
    let backButton = UIButton()
    let cellId = "cellId"

    let citiesCollectionView: UICollectionView = {
        let layout = UICollectionViewFlowLayout()
        let collection = UICollectionView(frame: CGRect.zero, collectionViewLayout: layout)
        layout.scrollDirection = .vertical
        layout.minimumInteritemSpacing = 0
        layout.minimumLineSpacing = 10
        return collection
    }()

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .white
        setup()
        layout()
    }

// nearly all of the code you currently have minus states or cities specific stuff
// ... etc.,
}
```

```swift
class StatesCollectionViewController: BTCollectionViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        self.mainTitle.text = "States"
        populate()
    }

    func populate() {
        fireBaseProxy.StatesRequest { (result) in
            // locations =  success value
            collectionView.reloadData()
        }
    }

    override cellForRowAt.....() {
        // Controller Specific Logic
    }
}

```

If you did that, you could literally have those three functions create as many new controllers as you want with the same format.

**Comment by JZDesign** [comment]
For example:

``` swift

class Button: UIButton {
    var action: () -> () = {}

    init(textColor: UIColor, backgroundColor: UIColor, title: String, cornerRadius: CGFloat = 0) {
        textColor = textColor
        backgroundColor = .....
        ......... etc.,
        addTarget(self, selector: #selector(buttonTapped)
    }

    func layoutIn(context: UIViewController) {
        button.top(context.SomeConstraint)....
    }

    @objc func buttonTapped() {
        action()
    }
}
```

```swift
let button = Button(.red, .white, "Title", 8)
button.action = { print("YAAAY") }
button.layoutIn(self)

...
let button2 = Button(.blue, .white, "Title2", 8)
```

**Comment by peytonshetler** [comment]
Very cool!!  Do you want me to refactor it?

**Comment by JZDesign** [comment]
I’ll leave that up to you. If you want to do that learning and refactor it I think you’d be better for it. But I’ll approve this so you can move forward if you want.

**Pull request merged**